gateway/src/node/client_handling/websocket/connection_handler/ecash/credential_sender.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright 2022-2024 - Nym Technologies SA <contact@nymtech.net>
+// SPDX-License-Identifier: GPL-3.0-only
+
+use std::sync::Arc;
+
+use crate::node::client_handling::websocket::connection_handler::authenticated::RequestHandlingError;
+use crate::node::storage::Storage;
+
+use futures::channel::mpsc::UnboundedReceiver;
+use futures::StreamExt;
+use nym_api_requests::coconut::models::VerifyEcashCredentialResponse;
+use nym_api_requests::coconut::VerifyEcashCredentialBody;
+
+use nym_gateway_requests::models::CredentialSpendingRequest;
+use nym_validator_client::nyxd::contract_traits::{EcashSigningClient, MultisigQueryClient};
+use nym_validator_client::nyxd::cosmwasm_client::logs::{find_attribute, BANDWIDTH_PROPOSAL_ID};
+use nym_validator_client::NymApiClient;
+use nym_validator_client::{nyxd::AccountId, DirectSigningHttpRpcNyxdClient};
+use tokio::{
+    sync::RwLock,
+    time::{interval, Duration},
+};
+
+const CRED_SENDING_INTERVAL: u64 = 300;
+
+#[derive(Clone)]
+pub struct PendingCredential {
+    pub credential: CredentialSpendingRequest,
+    pub address: AccountId,
+    pub api_clients: Vec<NymApiClient>,
+    pub proposal_id: Option<u64>,
+}
+
+impl PendingCredential {
+    pub fn new(
+        credential: CredentialSpendingRequest,
+        address: AccountId,
+        api_clients: Vec<NymApiClient>,
+    ) -> Self {
+        PendingCredential {
+            credential,
+            address,
+            api_clients,
+            proposal_id: None,
+        }
+    }
+}
+
+pub(crate) struct CredentialSender<St: Storage> {
+    cred_receiver: UnboundedReceiver<PendingCredential>,
+    storage: St,
+    nyxd_client: Arc<RwLock<DirectSigningHttpRpcNyxdClient>>,
+}
+
+impl<St> CredentialSender<St>
+where
+    St: Storage + 'static,
+{
+    pub(crate) fn new(
+        cred_receiver: UnboundedReceiver<PendingCredential>,
+        storage: St,
+        nyxd_client: Arc<RwLock<DirectSigningHttpRpcNyxdClient>>,
+    ) -> Self {
+        CredentialSender {
+            cred_receiver,
+            storage,
+            nyxd_client,
+        }
+    }
+
+    async fn create_proposal(
```

Comment:
this code is extremely similar to the one in `pub async fn spend_online_credential(`, could we somehow extract the shared code?

Reply:
Probably yes

Reply:
TODO after merge